src/chunks/vocab.rs

```diff
@@ -301,14 +294,90 @@ where
             .write_u32::<LittleEndian>(self.indexer.buckets() as u32)
             .map_err(|e| ErrorKind::io_error("Cannot write number of buckets", e))?;
 
-        for word in self.words() {
-            write
-                .write_u32::<LittleEndian>(word.len() as u32)
-                .map_err(|e| ErrorKind::io_error("Cannot write token length", e))?;
-            write
-                .write_all(word.as_bytes())
-                .map_err(|e| ErrorKind::io_error("Cannot write token", e))?;
-        }
+        write_vocab_items(write, self.words())?;
+
+        Ok(())
+    }
+}
+
+impl SubwordVocab<NGramIndexer> {
+    fn read_ngram_chunk<R>(
+        read: &mut R,
+        chunk_identifier: ChunkIdentifier,
+    ) -> Result<SubwordVocab<NGramIndexer>>
+    where
+        R: Read + Seek,
+    {
+        ChunkIdentifier::ensure_chunk_type(read, chunk_identifier)?;
+        // Read and discard chunk length.
+        read.read_u64::<LittleEndian>()
+            .map_err(|e| ErrorKind::io_error("Cannot read vocabulary chunk length", e))?;
+        let words_len = read
+            .read_u64::<LittleEndian>()
+            .map_err(|e| ErrorKind::io_error("Cannot read number of words", e))?;
+        let ngrams_len = read
+            .read_u64::<LittleEndian>()
+            .map_err(|e| ErrorKind::io_error("Cannot read number of ngrams", e))?;
+        let min_n = read
+            .read_u32::<LittleEndian>()
+            .map_err(|e| ErrorKind::io_error("Cannot read minimum n-gram length", e))?;
+        let max_n = read
+            .read_u32::<LittleEndian>()
+            .map_err(|e| ErrorKind::io_error("Cannot read maximum n-gram length", e))?;
+
+        let words = read_vocab_items(read, words_len as usize)?;
+        let ngrams = read_vocab_items(read, ngrams_len as usize)?;
+        let indexer = NGramIndexer::new(ngrams);
```

Comment:
Just a question: Is this variable ```indexer``` unnecessary or it doesn't matter?

Reply:
Thanks for the review! 

The `indexer` is used to retrieve the ngram indices in the `SubwordVocab`. For the other vocabularies, the indexer hashes an ngram and returns the hashcode, the hashcode does not necessarily map to a unique ngram, so there could be collisions. This indexer has a `HashTable` explicitly storing ngram-index mappings so that each ngram is assigned a unique index.

Reply:
Thanks for the answer! Sorry I didn't make it clear. Actually I mean in cases like this, I always do something like:
```
Ok(SubwordVocab::new(
    words, 
    min_n, 
    max_n, 
    indexer: NGramIndexer::new(ngrams),
)
```
Instead of creating another variable ```indexer```, and I'm not so sure if this is a common rule or it just doesn't matter that much :)

Reply:
I see, the two variants are equivalent, so it's just down to readability. I think both versions are commonly used, but personally I like to only use variables in the initialization and handle construction separately.